graph/graph.go

```diff
@@ -415,6 +417,42 @@ func (graph *Graph) ByParent() (map[string][]*image.Image, error) {
 	return byParent, err
 }
 
+// If the images and layers are in pulling chain, retain them.
+// If not, they may be deleted by rmi with dangling condition.
+func (graph *Graph) Retain(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.retained[layerID]; exists {
+			graph.retained[layerID] = append(imgs, imgID)
+		} else {
+			graph.retained[layerID] = []string{imgID}
+		}
+	}
+}
+
+// If the images and layers have been pulled, unretain them.
+func (graph *Graph) Unretain(layerIDs []string, imgID string) {
```

Comment:
I'm sorry but I found this loop to be a little awkward. Double check that this is doing what you want, but something like this is more readable:

```go
// Unretain removes the referenced image id from the provided set of layers.
func (graph *Graph) Unretain(layerIDs []string, imgID string) {
	for _, layerID := range layerIDs {
		imgIDs, ok := graph.retained[layerID]
		if !ok {
			continue // no references for this layerID
		}

		var filtered []string
		for i, candidate := range imgIDs {
			if candidate == imgID {
				continue // skip matching ids
			}

			filtered = append(filtered, candidate)
		}

		if len(filtered) > 0 {
			graph.retained[layerID] = filtered
		} else {
			delete(graph.retained, layerID) // delete empty reference slice
		}
	}
} 
```